File: bnc/src/main/java/org/sqlbuilder/bnc/BNCResolvingProcessor.java
package org.sqlbuilder.bnc;

import org.sqlbuilder.bnc.objects.BNCExtendedRecord;
import org.sqlbuilder.bnc.objects.BNCRecord;
import org.sqlbuilder.common.NotFoundException;
import org.sqlbuilder.common.ThrowingBiConsumer;
import org.sqlbuilder.common.Utils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.util.Properties;

public class BNCResolvingProcessor extends BNCProcessor
{
	protected final String serFile;

	protected final BncWordResolver wordResolver;

	public BNCResolvingProcessor(final Properties conf) throws IOException, ClassNotFoundException
	{
		super(conf);

		// output
		this.outDir = new File(conf.getProperty("bnc_outdir_resolved", "sql/data_resolved"));
		if (!this.outDir.exists())
		{
			this.outDir.mkdirs();
		}

		// resolve
		this.serFile = conf.getProperty("word_nids");
		this.wordResolver = new BncWordResolver(this.serFile);
	}

	@Override
	public void run() throws IOException
	{
		// main
		String bNCMain = conf.getProperty("bnc_main", "bnc.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.file("bncs"))), true, StandardCharsets.UTF_8))
		{
			processBNCFile(ps, new File(bncHome, bNCMain), names.table("bncs"), names.columns("bncs", true), (record, i) -> resolveAndInsert(ps, record, i));
		}

		// subfiles
		String bNCSpWr = conf.getProperty("bnc_spwr", "bnc-spoken-written.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.file("spwrs"))), true, StandardCharsets.UTF_8))
		{
			processBNCSubFile(ps, new File(bncHome, bNCSpWr), names.table("spwrs"), names.columns("spwrs", true), (record, i) -> resolveAndInsert(ps, record, i));
		}

		String bNCConvTask = conf.getProperty("bnc_convtask", "bnc-conv-task.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.file("convtasks"))), true, StandardCharsets.UTF_8))
		{
			processBNCSubFile(ps, new File(bncHome, bNCConvTask), names.table("convtasks"), names.columns("convtasks", true), (record, i) -> resolveAndInsert(ps, record, i));
		}

		String bNCImagInf = conf.getProperty("bnc_imaginf", "bnc-imag-inf.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.file("imaginfs"))), true, StandardCharsets.UTF_8))
		{
			processBNCSubFile(ps, new File(bncHome, bNCImagInf), names.table("imaginfs"), names.columns("imaginfs", true), (record, i) -> resolveAndInsert(ps, record, i));
		}
	}

	private void resolveAndInsert(final PrintStream ps, final BNCRecord record, final int i) throws NotFoundException
	{
		var nr = record.dataRow();
		var r = wordResolver.apply(record.word);
		if (r != null)
		{
			var values = String.format("%s,%s", nr, Utils.nullableInt(r));
			insertRow(ps, i, values);
			return;
		}
		throw new NotFoundException(record.word);
	}

	@Override
	protected void processBNCFile(final PrintStream ps, final File file, final String table, final String columns, final ThrowingBiConsumer<BNCRecord, Integer> consumer) throws IOException
	{
		ps.printf("INSERT INTO %s (%s) VALUES%n", table, columns);
		process(file, BNCRecord::parse, consumer);
		ps.print(';');
	}

	@Override
	protected void processBNCSubFile(final PrintStream ps, final File file, final String table, final String columns, final ThrowingBiConsumer<BNCRecord, Integer> consumer) throws IOException
	{
		ps.printf("INSERT INTO %s (%s) VALUES%n", table, columns);
		process(file, BNCExtendedRecord::parse, consumer);
		ps.print(';');
	}
}

File: bnc/src/main/java/org/sqlbuilder/bnc/BncResolvingProcessor.kt
package org.sqlbuilder.bnc

import org.sqlbuilder.bnc.objects.BncExtendedRecord
import org.sqlbuilder.bnc.objects.BncRecord
import org.sqlbuilder.common.NotFoundException
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.PrintStream
import java.nio.charset.StandardCharsets
import java.util.*

open class BncResolvingProcessor(conf: Properties) : BncProcessor(conf) {

    protected val serFile: String

    @JvmField
    protected val wordResolver: BncWordResolver

    init {
        // header
        header += "\n-- ${conf.getProperty("wn_resolve_against")}"

        // output
        outDir = File(conf.getProperty("bnc_outdir_resolved", "sql/data_resolved"))
        if (!outDir.exists()) {
            outDir.mkdirs()
        }

        // resolve
        serFile = conf.getProperty("word_nids")
        wordResolver = BncWordResolver(serFile)
    }

    @Throws(IOException::class)
    override fun processBNCFile(ps: PrintStream, file: File, table: String, columns: String, consumer: (BncRecord, Int) -> Unit) {
        ps.println("INSERT INTO $table ($columns) VALUES")
        process(file, { BncRecord.Companion.parse(it) }, consumer)
        ps.print(';')
    }

    @Throws(IOException::class)
    override fun processBNCSubFile(ps: PrintStream, file: File, table: String, columns: String, consumer: (BncRecord, Int) -> Unit) {
        ps.println("INSERT INTO $table ($columns) VALUES")
        process(file, { BncExtendedRecord.Companion.parse(it) }, consumer)
        ps.print(';')
    }

    @Throws(IOException::class)
    override fun run() {
        // main
        val bNCMain = conf.getProperty("bnc_main", "bnc.txt")
        PrintStream(FileOutputStream(File(outDir, names.file("bncs"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCFile(it, File(bncHome, bNCMain), names.table("bncs"), names.columns("bncs", true), { record: BncRecord, i: Int -> resolveAndInsert(it, record, i) })
        }
        // subfiles
        val bNCSpWr = conf.getProperty("bnc_spwr", "bnc-spoken-written.txt")
        PrintStream(FileOutputStream(File(outDir, names.file("spwrs"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCSubFile(it, File(bncHome, bNCSpWr), names.table("spwrs"), names.columns("spwrs", true), { record: BncRecord, i: Int -> resolveAndInsert(it, record, i) })
        }
        val bNCConvTask = conf.getProperty("bnc_convtask", "bnc-conv-task.txt")
        PrintStream(FileOutputStream(File(outDir, names.file("convtasks"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCSubFile(it, File(bncHome, bNCConvTask), names.table("convtasks"), names.columns("convtasks", true), { record: BncRecord, i: Int -> resolveAndInsert(it, record, i) })
        }
        val bNCImagInf = conf.getProperty("bnc_imaginf", "bnc-imag-inf.txt")
        PrintStream(FileOutputStream(File(outDir, names.file("imaginfs"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCSubFile(it, File(bncHome, bNCImagInf), names.table("imaginfs"), names.columns("imaginfs", true), { record: BncRecord, i: Int -> resolveAndInsert(it, record, i) })
        }
    }

    @Throws(NotFoundException::class)
    private fun resolveAndInsert(ps: PrintStream, record: BncRecord, i: Int) {
        val nr = record.dataRow()
        val r: Int? = wordResolver.apply(record.word)
        if (r != null) {
            val values = "$nr,$r"
            insertRow(ps, i.toLong(), values)
            return
        }
        throw NotFoundException(record.word)
    }
}
